File: src/frontend/src/services/__tests__/programService.test.ts
import { getPrograms, getProgram, createProgram, reviewProgram } from '../programService';
import apiClient from '../apiClient';

vi.mock('../apiClient', () => ({
  default: {
    get: vi.fn(),
    post: vi.fn(),
    put: vi.fn(),
  },
}));

const mockedApiClient = apiClient as {
  get: ReturnType<typeof vi.fn>;
  post: ReturnType<typeof vi.fn>;
  put: ReturnType<typeof vi.fn>;
};

describe('programService', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('getPrograms', () => {
    it('returns paged programs from the API', async () => {
      const mockData = {
        content: [{ id: 1, programName: 'Test', status: 'SUBMITTED' }],
        totalElements: 1,
        totalPages: 1,
        page: 0,
        size: 20,
      };
      mockedApiClient.get.mockResolvedValue({ data: mockData });

      const result = await getPrograms({});
      expect(result).toEqual(mockData);
      expect(mockedApiClient.get).toHaveBeenCalledWith('/programs', expect.any(Object));
    });

    it('passes default page=0 and size=20', async () => {
      mockedApiClient.get.mockResolvedValue({ data: { content: [], totalElements: 0, totalPages: 0, page: 0, size: 20 } });
      await getPrograms({});
      expect(mockedApiClient.get).toHaveBeenCalledWith('/programs', {
        params: { status: undefined, programTypeId: undefined, page: 0, size: 20 },
      });
    });

    it('passes status and programTypeId filters when provided', async () => {
      mockedApiClient.get.mockResolvedValue({ data: { content: [], totalElements: 0, totalPages: 0, page: 0, size: 20 } });
      await getPrograms({ status: 'APPROVED', programTypeId: 2, page: 1, size: 10 });
      expect(mockedApiClient.get).toHaveBeenCalledWith('/programs', {
        params: { status: 'APPROVED', programTypeId: 2, page: 1, size: 10 },
      });
    });
  });

  describe('getProgram', () => {
    it('returns a single program by id', async () => {
      const mockProgram = { id: 42, programName: 'Test', status: 'DRAFT', programTypeId: 1, createdAt: '2026-01-01T00:00:00Z' };
      mockedApiClient.get.mockResolvedValue({ data: mockProgram });

      const result = await getProgram(42);
      expect(result).toEqual(mockProgram);
      expect(mockedApiClient.get).toHaveBeenCalledWith('/programs/42');
    });
  });

  describe('createProgram', () => {
    it('posts program data and returns created program', async () => {
      const request = { programName: 'New', programTypeId: 1, createdBy: 'a@b.com' };
      const mockProgram = { id: 5, programName: 'New', programTypeId: 1, status: 'SUBMITTED', createdAt: '2026-01-01T00:00:00Z', createdBy: 'a@b.com' };
      mockedApiClient.post.mockResolvedValue({ data: mockProgram });

      const result = await createProgram(request);
      expect(result).toEqual(mockProgram);
      expect(mockedApiClient.post).toHaveBeenCalledWith('/programs', request);
    });
  });

  describe('reviewProgram', () => {
    it('puts review data and returns updated program', async () => {
      const reviewData = { status: 'APPROVED' as const, reviewerComments: 'Good' };
      const mockProgram = { id: 1, programName: 'Test', status: 'APPROVED', programTypeId: 1, createdAt: '2026-01-01T00:00:00Z' };
      mockedApiClient.put.mockResolvedValue({ data: mockProgram });

      const result = await reviewProgram(1, reviewData);
      expect(result).toEqual(mockProgram);
      expect(mockedApiClient.put).toHaveBeenCalledWith('/programs/1/review', reviewData);
    });
  });
});

File: src/test/java/com/ontario/program/service/ProgramServiceTest.java
package com.ontario.program.service;

import com.ontario.program.config.ResourceNotFoundException;
import com.ontario.program.dto.CreateProgramRequest;
import com.ontario.program.dto.PagedResponse;
import com.ontario.program.dto.ProgramResponse;
import com.ontario.program.dto.ReviewProgramRequest;
import com.ontario.program.model.Program;
import com.ontario.program.model.ProgramType;
import com.ontario.program.repository.ProgramRepository;
import com.ontario.program.repository.ProgramTypeRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class ProgramServiceTest {

    @Mock
    private ProgramRepository programRepository;

    @Mock
    private ProgramTypeRepository programTypeRepository;

    @InjectMocks
    private ProgramService programService;

    private ProgramType testProgramType;
    private Program testProgram;

    @BeforeEach
    void setUp() {
        testProgramType = new ProgramType("Community Services", "Services communautaires");
        testProgramType.setId(1);

        testProgram = new Program();
        testProgram.setId(1);
        testProgram.setProgramName("Youth Employment Initiative");
        testProgram.setProgramDescription("A program to help youth find employment");
        testProgram.setProgramType(testProgramType);
        testProgram.setStatus("SUBMITTED");
        testProgram.setCreatedAt(LocalDateTime.now());
        testProgram.setCreatedBy("citizen@example.com");
    }

    @Test
    void givenProgramsExist_whenGetPrograms_thenReturnPagedResponse() {
        // Given
        Page<Program> programPage = new PageImpl<>(List.of(testProgram));
        when(programRepository.findAll(any(Pageable.class))).thenReturn(programPage);

        // When
        PagedResponse<ProgramResponse> result = programService.getPrograms(null, null, 0, 20);

        // Then
        assertThat(result.content()).hasSize(1);
        assertThat(result.content().get(0).programName()).isEqualTo("Youth Employment Initiative");
        assertThat(result.totalElements()).isEqualTo(1);
    }

    @Test
    void givenProgramExists_whenGetProgramById_thenReturnProgram() {
        // Given
        when(programRepository.findById(1)).thenReturn(Optional.of(testProgram));

        // When
        Optional<ProgramResponse> result = programService.getProgramById(1);

        // Then
        assertThat(result).isPresent();
        assertThat(result.get().programName()).isEqualTo("Youth Employment Initiative");
        assertThat(result.get().programTypeName()).isEqualTo("Community Services");
    }

    @Test
    void givenProgramNotExists_whenGetProgramById_thenReturnEmpty() {
        // Given
        when(programRepository.findById(999)).thenReturn(Optional.empty());

        // When
        Optional<ProgramResponse> result = programService.getProgramById(999);

        // Then
        assertThat(result).isEmpty();
    }

    @Test
    void givenValidRequest_whenCreateProgram_thenReturnCreatedProgram() {
        // Given
        CreateProgramRequest request = new CreateProgramRequest(
            "New Program", "Description", 1, "user@example.com"
        );
        when(programTypeRepository.findById(1)).thenReturn(Optional.of(testProgramType));
        when(programRepository.save(any(Program.class))).thenAnswer(invocation -> {
            Program saved = invocation.getArgument(0);
            saved.setId(2);
            return saved;
        });

        // When
        ProgramResponse result = programService.createProgram(request);

        // Then
        assertThat(result.id()).isEqualTo(2);
        assertThat(result.programName()).isEqualTo("New Program");
        assertThat(result.status()).isEqualTo("SUBMITTED");
    }

    @Test
    void givenInvalidProgramType_whenCreateProgram_thenThrowException() {
        // Given
        CreateProgramRequest request = new CreateProgramRequest(
            "New Program", "Description", 999, "user@example.com"
        );
        when(programTypeRepository.findById(999)).thenReturn(Optional.empty());

        // When/Then
        assertThatThrownBy(() -> programService.createProgram(request))
            .isInstanceOf(ResourceNotFoundException.class)
            .hasMessageContaining("Program type not found");
    }

    @Test
    void givenSubmittedProgram_whenReviewApprove_thenReturnApprovedProgram() {
        // Given
        testProgram.setStatus("SUBMITTED");
        ReviewProgramRequest request = new ReviewProgramRequest("APPROVED", "Looks good!");
        when(programRepository.findById(1)).thenReturn(Optional.of(testProgram));
        when(programRepository.save(any(Program.class))).thenAnswer(invocation -> invocation.getArgument(0));

        // When
        ProgramResponse result = programService.reviewProgram(1, request);

        // Then
        assertThat(result.status()).isEqualTo("APPROVED");
        assertThat(result.reviewerComments()).isEqualTo("Looks good!");
    }

    @Test
    void givenDraftProgram_whenReview_thenThrowException() {
        // Given
        testProgram.setStatus("DRAFT");
        ReviewProgramRequest request = new ReviewProgramRequest("APPROVED", "Comments");
        when(programRepository.findById(1)).thenReturn(Optional.of(testProgram));

        // When/Then
        assertThatThrownBy(() -> programService.reviewProgram(1, request))
            .isInstanceOf(IllegalStateException.class)
            .hasMessageContaining("Only SUBMITTED programs can be reviewed");
    }
}
